tsdb/exemplar.go

```diff
@@ -0,0 +1,202 @@
+		}),
+		duplicateExemplars: prometheus.NewCounter(prometheus.CounterOpts{
+			Name: "prometheus_exemplar_duplicate_exemplars_total",
+			Help: "Total number of series in the head block.",
+		}),
+	}
+	if r != nil {
+		r.MustRegister(
+			m.outOfOrderExemplars,
+			m.duplicateExemplars,
+		)
+	}
+	return m
+}
+
+type CircularExemplarStorage struct {
+	metrics   *exemplarMetrics
+	lock      sync.RWMutex
+	index     map[string]indexEntry
+	exemplars []*circularBufferEntry
+	nextIndex int
+}
+
+type indexEntry struct {
+	first int
+	last  int
+}
+
+type circularBufferEntry struct {
+	exemplar     exemplar.Exemplar
+	seriesLabels labels.Labels
+	next         int
+}
+
+// If we assume the average case 95 bytes per exemplar we can fit 5651272 exemplars in
+// 1GB of extra memory, accounting for the fact that this is heap allocated space.
+func NewCircularExemplarStorage(len int, reg prometheus.Registerer) (*CircularExemplarStorage, error) {
+	if len < 1 {
+		return nil, errors.Errorf("must initialize exemplar storage with space for at least 1 exemplar, %d is invalid", len)
+	}
+	return &CircularExemplarStorage{
+		exemplars: make([]*circularBufferEntry, len),
+		index:     make(map[string]indexEntry),
+		metrics:   newExemplarMetrics(reg),
+	}, nil
+}
+
+func (ce *CircularExemplarStorage) Appender() storage.ExemplarAppender {
+	return ce
+}
+
+func (ce *CircularExemplarStorage) ExemplarAppender() storage.ExemplarAppender {
+	return ce
+}
+
+func (ce *CircularExemplarStorage) ExemplarQuerier(_ context.Context) (storage.ExemplarQuerier, error) {
+	return ce, nil
+}
+
+func (ce *CircularExemplarStorage) Querier(ctx context.Context) (storage.ExemplarQuerier, error) {
+	return ce, nil
+}
+
+// Select returns exemplars for a given set of series labels hash.
+func (ce *CircularExemplarStorage) Select(start, end int64, l labels.Labels) ([]exemplar.Exemplar, error) {
+	var (
+		ret []exemplar.Exemplar
+		e   *circularBufferEntry
+		idx indexEntry
+		ok  bool
+	)
+
+	ce.lock.RLock()
+	defer ce.lock.RUnlock()
+
+	if idx, ok = ce.index[l.String()]; !ok {
+		return nil, nil
+	}
+
+	e = ce.exemplars[idx.first]
+	for {
+		if e.exemplar.Ts < start {
+			if e.next == -1 {
+				break
+			}
+			e = ce.exemplars[e.next]
+			continue
+		}
+		if e.exemplar.Ts > end {
+			break
+		}
+
+		ret = append(ret, e.exemplar)
+		if e.next == -1 {
+			break
+		}
+		e = ce.exemplars[e.next]
+	}
+	return ret, nil
+}
+
+// Takes the circularBufferEntry that will be overwritten and updates the
+// storages index for that entries labelset if necessary.
+func (ce *CircularExemplarStorage) indexGc(cbe *circularBufferEntry) {
+	if cbe == nil {
+		return
+	}
+
+	l := cbe.seriesLabels.String()
+	i := cbe.next
+	if i == -1 {
+		delete(ce.index, l)
+		return
+	}
+
+	ce.index[l] = indexEntry{i, ce.index[l].last}
+}
+
+func (ce *CircularExemplarStorage) AddExemplar(l labels.Labels, e exemplar.Exemplar) error {
+	seriesLabels := l.String()
+	ce.lock.Lock()
+	defer ce.lock.Unlock()
```

Comment:
It definitely could be a bottleneck. Not sure what we could do right now since exemplars for all series are stored in the same circular buffer.

I'll look at the current benchmark for sample ingestion.